### PR TITLE
Add avif image file support

### PR DESCRIPTION
**Pull request by wxiaoguang**
Most modern browsers support it now

` Update ALLOWED_TYPES #96 ` https://gitea.com/gitea/docs/pulls/96

### CODE REVIEW

File: custom/conf/app.example.ini [hunk 1]
```diff
@@ -1912,7 +1912,7 @@ LEVEL = Info
 ;ENABLED = true
 ;;
 ;; Comma-separated list of allowed file extensions (`.zip`), mime types (`text/plain`) or wildcard type (`image/*`, `audio/*`, `video/*`). Empty value or `*/*` allows all types.
-;ALLOWED_TYPES = .csv,.docx,.fodg,.fodp,.fods,.fodt,.gif,.gz,.jpeg,.jpg,.log,.md,.mov,.mp4,.odf,.odg,.odp,.ods,.odt,.patch,.pdf,.png,.pptx,.svg,.tgz,.txt,.webm,.xls,.xlsx,.zip
+;ALLOWED_TYPES = .avif,.csv,.docx,.fodg,.fodp,.fods,.fodt,.gif,.gz,.jpeg,.jpg,.log,.md,.mov,.mp4,.odf,.odg,.odp,.ods,.odt,.patch,.pdf,.png,.pptx,.svg,.tgz,.txt,.webm,.webp,.xls,.xlsx,.zip
```
Comment (silverwind) [inline custom/conf/app.example.ini]:
```suggestion
;ALLOWED_TYPES = .avif,.cpuprofile,.csv,.dmp,.docx,.fodg,.fodp,.fods,.fodt,.gif,.gz,.jpeg,.jpg,.json,.jsonc,.log,.md,.mov,.mp4,.odf,.odg,.odp,.ods,.odt,.patch,.pdf,.png,.pptx,.svg,.tgz,.txt,.webm,.webp,.xls,.xlsx,.zip
```

File: modules/setting/attachment.go [hunk 2]
```diff
@@ -3,33 +3,33 @@
...
-	MaxSize:      2048,
-	MaxFiles:     5,
-	Enabled:      true,
 }
 
+var Attachment AttachmentSettingType
+
 func loadAttachmentFrom(rootCfg ConfigProvider) (err error) {
+	Attachment = AttachmentSettingType{
+		AllowedTypes: ".avif,.cpuprofile,.csv,.dmp,.docx,.fodg,.fodp,.fods,.fodt,.gif,.gz,.jpeg,.jpg,.json,.jsonc,.log,.md,.mov,.mp4,.odf,.odg,.odp,.ods,.odt,.patch,.pdf,.png,.pptx,.svg,.tgz,.txt,.webm,.webp,.xls,.xlsx,.zip",
```
Comment (silverwind) [inline modules/setting/attachment.go:18]:
```suggestion
		AllowedTypes: ".avif,.cpuprofile,.csv,.dmp,.docx,.fodg,.fodp,.fods,.fodt,.gif,.gz,.jpeg,.jpg,.json,.jsonc,.log,.md,.mov,.mp4,.odf,.odg,.odp,.ods,.odt,.patch,.pdf,.png,.pptx,.svg,.tgz,.txt,.webm,.webp,.xls,.xlsx,.zip",
```

Comment (wxiaoguang) [reply modules/setting/attachment.go:18]: What's the difference?

Comment (silverwind) [reply modules/setting/attachment.go:18]: Aparrently none, I just copied from the config file.

File: modules/typesniffer/typesniffer.go [hunk 3]
```diff
@@ -81,6 +83,25 @@ func (ct SniffedType) GetMimeType() string {
...
+	tag := string(data[4:8])
+	if tag != "ftyp" {
+		return nil, false
+	}
+	brands = append(brands, string(data[8:12]))
+	for i := 16; i+4 <= boxSize; i += 4 {
+		brands = append(brands, string(data[i:i+4]))
+	}
+	return brands, true
+}
```
Comment (silverwind) [inline modules/typesniffer/typesniffer.go:104]: Some comment what this does would be helpful. Magic byte detection?

**Comment by wxiaoguang** [comment]
The list in code is up-to-dated. Some items are missing in the app.example.ini due to #29688 .......

File: modules/typesniffer/typesniffer.go (hunk 3 above)
Comment (wxiaoguang) [reply modules/typesniffer/typesniffer.go:104]: I think the code is clear enough, it is called file type box (ftyp)

Comment (silverwind) [reply modules/typesniffer/typesniffer.go:104]: Add a link to some docs about it?

Comment (wxiaoguang) [reply modules/typesniffer/typesniffer.go:104]: google it ,  a lot : https://developer.apple.com/documentation/quicktime-file-format/file_type_compatibility_atom

But I guess we won't like to put an apple link here.

**Comment by silverwind** [comment]
> The list in code is up-to-dated. Some items are missing in the app.example.ini due to #29688 .......

Not sure how the mismatch came to be. What there an error in #29688?

**Comment by wxiaoguang** [comment]
> > The list in code is up-to-dated. Some items are missing in the app.example.ini due to #29688 .......
> 
> Not sure how the mismatch came to be. What there an error in #29688?

app.example.ini wasn't updated there

File: modules/typesniffer/typesniffer.go (hunk 3 above)
Comment (silverwind) [reply modules/typesniffer/typesniffer.go:104]: How about a wikipedia link?

https://en.wikipedia.org/wiki/ISO_base_media_file_format#File_type_box

Comment (wxiaoguang) [reply modules/typesniffer/typesniffer.go:104]: And by the way, the AVIF related ISO standards are not free, it needs to pay money to download, I didn't get one copy, so I just collected some information from internet.

**Comment by silverwind** [comment]
Ah, I see. Yes, that was an error.

File: modules/typesniffer/typesniffer.go [hunk 4]
```diff
@@ -81,6 +83,25 @@ func (ct SniffedType) GetMimeType() string {
 	return strings.SplitN(ct.contentType, ";", 2)[0]
 }
 
+func detectFileTypeBox(data []byte) (brands []string, found bool) {
```
Comment (silverwind) [inline modules/typesniffer/typesniffer.go:87]:
```suggestion
// https://en.wikipedia.org/wiki/ISO_base_media_file_format#File_type_box
func detectFileTypeBox(data []byte) (brands []string, found bool) {
```

**Review by silverwind — APPROVED** [review]

**Comment by silverwind** [comment]
Optional, but might be good to verify whether avif files can be used as avatar. I assume it will need some changes.

**Review by lunny — APPROVED** [review]

**Pull request merged**